flow: Make the `Actions` type work + various cleanups

roberthoenig opened the pull request:
Originally, this set out to get flow into a state where it will complain about bugs like the one fixed in https://github.com/zulip/zulip-mobile/commit/d5e6a0bb2c66b793155d05a6b10be57bcdeb6a5a.
The issue here was the following `dispatch` call:
```js
dispatch(navigateToCreateStream);
```
`navigateToCreateStream` is an action creator, so it needs to be called before it can be dispatched. The correct `dispatch` call is
```js
dispatch(navigateToCreateStream());
```

I only succeeded partially. The dispatch bug described above seems to be in fact undetectable with the current implementation of Flow. On the other hand, the commits in this PR make Flow detect some other potential regression candidates, among some other things.

Why can't Flow detect the dispatch bug described above?

I tried to approach the bug by looking at the type of `dispatch`. It's probably too permissive, since it accepts an action creator, even though it should only accept actions. The type of `dispatch` is `Dispatch`:
```js
export type Thunk<A> = ((Dispatch, GetState) => Promise<any> | void) => A;
export type Dispatch = ReduxDispatch<Action> & Thunk<Action>;
```
`ReduxDispatch<Action>` is `Action` object.  With this PR, the `Action` type works properly (and doesn't accept anything anymore), so there must  be another issue.  The issue lies with `Thunk<A>`: This allows actions to be functions, and not just objects. In particular, an action can be the `(Dispatch, GetState) => Promise<any> | void`. Now,  `navigateToCreateStream` has an arity of 0, whereas Flow should expect a function of arity 2, right? Nope! In Flow, the type of a function that accepts only a subset of the parameters of another function is a subtype of the type of that function; see the [Flow function subtyping rules](https://flow.org/en/docs/lang/subtypes/#toc-subtypes-of-functions).

This is very annoying, and there is no way to work around this by e.g. declaring "exact functions" that require a certain arity. The workaround I can think of is very annoying as well: We could split the Dispatch type, but that wouldn't reflect reality and create a ton of new boilerplate annotations.

I guess for this case our best bet is living with this annoyance, but adding a linter rule so that the specific dispatch regression would be caught.

borisyankov commented on src/types.js:
```diff
@@ -13,7 +13,7 @@ export type MapStateToProps = any; // { MapStateToProps } from 'react-redux';
 export type * from './api/apiTypes';
 
-export type Thunk<A> = ((Dispatch, GetState) => Promise<void> | void) => A;
+export type Thunk<A> = ((Dispatch, GetState) => Promise<any> | void) => A;
```
I think we should keep it `void` and not return values from these actions?

Just adding `{` and `}` around toggleStreamNotification's body and turning it to `Promise<void>` seems to me to match our intent better.

roberthoenig replied:
Yeah, I tried that initially. However, toggleStreamNotification is an `async` function, so it'll always return a promise. And that promise will not have the type void.

borisyankov replied:
It will if you add the `{` and `}` around the function. Right now it is returning the result of `apiPost` and we are not using that.

roberthoenig replied:
Aah, good point! I looked at toggleStreamNotification*s*, didn't think about toggleStreamNotification

borisyankov commented on src/session/sessionActions.js:
```diff
@@ -37,11 +30,7 @@ export const appOnline = (isOnline: boolean) => (
-  getState: GetState,
-  // $FlowFixMe
-): ?AppStateAction => {
+export const appState = (isActive: boolean) => (dispatch: Dispatch, getState: GetState): void => {
```
The type of the first function will be something like `Thunk<AppStateAction>` (if we imagine we have declared `Thunk` already)

roberthoenig replied:
Do you mean we should annotate this in a new commit?

roberthoenig commented:
@borisyankov updated to address your comment
> I think we should keep it void and not return values from these actions?

gnprice commented:
Thanks @roberthoenig for this detailed investigation and fixes! I am excited to see this get straightened out so that we get better type information, and catch more errors so we can do more refactors with more confidence. :-)

I've just merged the first few of these commits, adding all the explicit action types. I've held off on the commits that remove the fixmes, because I think there may actually be a cleaner fix by fixing the type we write down as `Dispatch` -- [see chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/subject/Redux.20types/near/600769).

I think this fix still won't get us to your original goal, unfortunately -- the issue is that an action creator we've forgotten to call looks, as far as its type is concerned, exactly like a thunk action that for some reason makes no use of its arguments.

gnprice commented:
I also went and merged the commit to turn on warnings, and remove the unnecessary fixme comments.

For the change to rewrite `any` as `$FlowFixMe`: note there are a fair number of other `any` annotations in the codebase! Probably most of them should also be replaced with real types -- maybe even all, as there's `mixed` and other things and I think the idea is supposed to be that `any` is never needed in good codebase once you've fully adopted types.

I'd be on board with replacing these with some notation that says more explicitly that we really mean to replace them with real types -- especially the ones like in this commit where we understand the situation and we know that's true. I think I'd want it to be something different from the notation we use in places like this commit series, where we've just uncovered a type error because some specific types in two places clash when we fill in correct types for something else that's between them. The latter feels different to me, in particular that it calls for some explanation; and I like being able to grep for it specifically.

We can add more patterns in `.flowconfig`, whatever we like. Maybe `$FlowAddMe`?

roberthoenig commented:
Updated:

* Removed the commit for rewriting `any` as `$FlowFixMe`. If we feel that most-all `any`s should be replaced, then there might not be much sense in replacing them with $Flow* at all. Also, this commit was half-hearted and left many `any`s intact. This could confuse people into thinking that those `any`s are wanted.

* Added the `Dispatch` type fix.
  > 've held off on the commits that remove the fixmes, because I think there may actually be a cleaner fix by fixing the type we write down as Dispatch
  
  I still think that for appOnline etc. the type should be `void` instead of `?App*Action`. I don't see how it's possible for these function to ever return anything. 

* Added a flow type test suite.

gnprice commented:
Thanks -- merged! Without the "DON'T MERGE ME" test-testing commit. ;-)

I also added another commit on top to replace `ReduxDispatch<Action>` with `ReduxDispatch<*>`, which allowed removing another fixme.

> I still think that for appOnline etc. the type should be void instead of ?App*Action. I don't see how it's possible for these function to ever return anything.

Ah, indeed -- I agree. I was thinking of the change that added braces.
